test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/regex/RegexMatcherTest.scala

```diff
@@ -33,42 +33,129 @@ class RegexMatcherTest  {
+    val matcher = Pattern.compile("\\s(([A-Za-z]{5}(hum)?).js)\\s").matcher(
+      "Write Scala.js everyday!")
 
-    def checkGroup0(start: Int, end: Int, group: String): Unit =
-      checkGroup(start, 5, end, 15, group, " Scala.js ")
+    assertTrue(matcher.find())
+    assertEquals(3, matcher.groupCount)
+    checkGroups(matcher,
+      (5, 15, " Scala.js "),
+      (6, 14, "Scala.js"),
+      (6, 11, "Scala"),
+      (-1, -1, null)
+    )
+  }
 
-    def checkGroup1(start: Int, end: Int, group: String): Unit =
-      checkGroup(start, 6, end, 14, group, "Scala.js")
+  @Test def start_end_group_tricky_and_toMatchResult(): Unit = {
+    val matcher = Pattern.compile("(Scala\\.js).*(Scala)$").matcher(
+      "Scala.js is a Scalable javascript compiler based on Scala")
 
-    def checkGroup2(start: Int, end: Int, group: String): Unit =
-      checkGroup(start, 6, end, 11, group, "Scala")
+    assertTrue(matcher.find())
+    assertEquals(2, matcher.groupCount)
+    checkGroups(matcher,
+      (0, 57, "Scala.js is a Scalable javascript compiler based on Scala"),
+      (0, 8, "Scala.js"),
+      (52, 57, "Scala")
+    )
+  }
 
-    def checkGroup3(start: Int, end: Int, group: String): Unit =
-      checkGroup(start, -1, end, -1, group, null)
+  @Test def start_end_group_matchnot_and_toMatchResult(): Unit = {
+    val matcher = Pattern.compile("(?!Scala\\.js)(Scala)").matcher(
+      "There is a difference between Scala.js and Scala, but both are Scalable")
 
-    def checkGroup(start: Int, startExpected: Int, end: Int, endExpected: Int,
-                   group: String, groupExpected: String): Unit = {
-      assertEquals(startExpected, start)
-      assertEquals(endExpected, end)
-      assertEquals(groupExpected, group)
-    }
+    assertTrue(matcher.find())
+    assertEquals(1, matcher.groupCount)
+    checkGroups(matcher,
+      (43, 48, "Scala"),
+      (43, 48, "Scala")
+    )
+    assertTrue(matcher.find())
+    checkGroups(matcher,
+      (63, 68, "Scala"),
+      (63, 68, "Scala")
+    )
+  }
+
+  @Test def start_end_group_multiple_and_toMatchResult(): Unit = {
+    val matcher = Pattern.compile("(?=Scala\\.js is (nice|awesome))(Scala)\\.js").matcher(
+      "Scala.js is nice, Scala.js is awesome")
 
     assertTrue(matcher.find())
-    assertEquals(3, matcher.groupCount)
-    checkGroup0(matcher.start, matcher.end, matcher.group)
-    checkGroup0(matcher.start(0), matcher.end(0), matcher.group(0))
-    checkGroup1(matcher.start(1), matcher.end(1), matcher.group(1))
-    checkGroup2(matcher.start(2), matcher.end(2), matcher.group(2))
-    checkGroup3(matcher.start(3), matcher.end(3), matcher.group(3))
+    checkGroups(matcher,
+      (0,  8,  "Scala.js"),
+      (12, 16, "nice"),
+      (0,  5,  "Scala")
+    )
+    assertTrue(matcher.find())
+    checkGroups(matcher,
+      (18, 26, "Scala.js"),
+      (30, 37, "awesome"),
+      (18, 23, "Scala")
+    )
+  }
 
+  def parseExpect(regex: String, str: String, pos: Int*): Unit = {
```

Comment:
This would be much clearer if it took a `startsAndEnds: (Int, Int)*`. It would make the call sites much more readable as they would be written as `(0, 2), (1, 2)` instead of `0, 2, 1, 2`. Or even `0 -> 2, 1 -> 2`, which is arguably a nice DSL for `start -> end`.

Reply:
New DSL Start -> End implemented.